### PR TITLE
[AArch64][PAC] Move emission of LR checks in tail calls to AsmPrinter

### PR DESCRIPTION
**Pull request by atrosinenko**
Move the emission of the checks performed on the authenticated LR value
during tail calls to AArch64AsmPrinter class, so that different checker
sequences can be reused by pseudo instructions expanded there.
This adds one more option to AuthCheckMethod enumeration, the generic
XPAC variant which is not restricted to checking the LR register.

### CODE REVIEW

**Comment by atrosinenko** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/110705?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#110705** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/110705?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#110702** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/110702?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @atrosinenko and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by atrosinenko** [comment]
Updated the commit message: removed the paragraph about dropping one `mov` instruction from the non-trapping variant of check. That was my initial idea to make non-hint `xpac(i|d)` operate on the tested register itself (just like `xpaclri` does), but it was removed from the final version of the patch, to not make unnecessary changes to the tests.

**Comment by kovdan01** [comment]
@atrosinenko With this patch applied, the following tests in test-suite compiled in Release mode become failed with segmentation fault:

1. SingleSource/Benchmarks/Misc-C++-EH/spirit.test
2. MultiSource/Benchmarks/tramp3d-v4/tramp3d-v4.test
3. MultiSource/Benchmarks/Prolangs-C++/shapes/shapes.test
4. MultiSource/Benchmarks/Prolangs-C++/ocean/ocean.test
5. MultiSource/Benchmarks/Bullet/bullet.test
6. MultiSource/Applications/lambda-0.1.3/lambda.test
7. MultiSource/Applications/kimwitu++/kc.test

I'm now preparing a minimal reproducer for that. Stay tuned.

**Comment by kovdan01** [comment]
@atrosinenko Here is a reproducer IR code.

<details>

```
define void @_ZN7myshape4moveEv(ptr %this) #0 {
entry:
  %0 = load ptr, ptr %this, align 8
  %vtable = load ptr, ptr %0, align 8
  %1 = ptrtoint ptr %0 to i64
  %2 = tail call i64 @llvm.ptrauth.blend(i64 %1, i64 50297)
  %3 = ptrtoint ptr %vtable to i64
  %4 = tail call i64 @llvm.ptrauth.auth(i64 %3, i32 2, i64 %2)
  %5 = inttoptr i64 %4 to ptr
  %6 = load ptr, ptr %5, align 8
  %7 = tail call i64 @llvm.ptrauth.blend(i64 %4, i64 36564)
  tail call void %6(ptr noundef nonnull align 8 dereferenceable(8) %0) [ "ptrauth"(i32 0, i64 %7) ]
  %r = getelementptr inbounds i8, ptr %this, i64 8
  %8 = load ptr, ptr %r, align 8
  %vtable2 = load ptr, ptr %8, align 8
  %9 = ptrtoint ptr %8 to i64
  %10 = tail call i64 @llvm.ptrauth.blend(i64 %9, i64 50297)
  %11 = ptrtoint ptr %vtable2 to i64
  %12 = tail call i64 @llvm.ptrauth.auth(i64 %11, i32 2, i64 %10)
  %13 = inttoptr i64 %12 to ptr
  %14 = load ptr, ptr %13, align 8
  %15 = tail call i64 @llvm.ptrauth.blend(i64 %12, i64 36564)
  tail call void %14(ptr noundef nonnull align 8 dereferenceable(8) %8) [ "ptrauth"(i32 0, i64 %15) ]
  ret void
}

declare i64 @llvm.ptrauth.blend(i64, i64)

declare i64 @llvm.ptrauth.auth(i64, i32 immarg, i64)

attributes #0 = { "ptrauth-auth-traps" "ptrauth-calls" "ptrauth-returns" }
```

</details>

When compiled with `llc -O3 -mtriple aarch64-linux-gnu -mattr=+pauth`, the following assembly is generated. X16 register is clobbered during check inserted under `else if (Method == AuthCheckMethod::HighBitsNoTBI)`. The previous value (containing pointer for address discrimination) is not spilled before and not restored after that. So, the last `braa` instructio uses incorrect discrimination value in X16 register, which results in authentication failure and trying to pass control to an address with non-zero pac field, causing a segfault. nProbably, a different scratch register (e.g. X17) should be used for check sequence if X16 is already in use for address discrimination.

<details>

```
_ZN7myshape4moveEv:                     // @_ZN7myshape4moveEv
	.cfi_startproc
// %bb.0:                               // %entry
	.cfi_b_key_frame
	pacibsp
	stp	x30, x19, [sp, #-16]!           // 16-byte Folded Spill
	.cfi_negate_ra_state
	.cfi_def_cfa_offset 16
	.cfi_offset w19, -8
	.cfi_offset w30, -16
	mov	x19, x0
	ldr	x0, [x0]
	ldr	x16, [x0]
	mov	x17, x0
	movk	x17, #50297, lsl #48
	autda	x16, x17
	mov	x17, x16
	xpacd	x17
	cmp	x16, x17
	b.eq	.Lauth_success_0
	brk	#0xc472
.Lauth_success_0:
	ldr	x9, [x16]
	mov	x8, x16
	mov	x17, x8
	movk	x17, #36564, lsl #48
	blraa	x9, x17
	ldr	x0, [x19, #8]
	ldr	x16, [x0]
	mov	x17, x0
	movk	x17, #50297, lsl #48
	autda	x16, x17
	mov	x17, x16
	xpacd	x17
	cmp	x16, x17
	b.eq	.Lauth_success_1
	brk	#0xc472
.Lauth_success_1:
	ldr	x1, [x16]
	ldp	x30, x19, [sp], #16             // 16-byte Folded Reload
	autibsp
	eor	x16, x30, x30, lsl #1
	tbz	x16, #62, .Lauth_success_2
	brk	#0xc471
.Lauth_success_2:
	movk	x16, #36564, lsl #48
	braa	x1, x16
```

</details>

The IR corresponds to the following C++ code.

<details>

```
struct shape {
  virtual void move() = 0;
};

struct line : public shape {
  void move () override;
};

struct myshape {
  line *l;
  line *r;
  void move();
};

void myshape::move() {
  l->move();
  r->move();
}
```

</details>

**Review by kovdan01 — DISMISSED** [review]
See https://github.com/llvm/llvm-project/pull/110705#issuecomment-2396047508

**Comment by atrosinenko** [comment]
@kovdan01 Thank you for the reproducer!

I restored the original check `TI->readsRegister(AArch64::X16, TRI) ? AArch64::X17 : AArch64::X16` when computing the scratch register, as checking just the specific operand did not handle the `AUTH_TCRETURN*` pseudos which actually have two register operands. Additionally, the register class of the second register operand of these pseudo instructions was restricted, as it seems to be just a coincidence that at least one scratch register was available for authenticated tail calls.

**Comment by atrosinenko** [comment]
Considering the reproducer, the code became a bit longer but correct:

<spoiler>
```
_ZN7myshape4moveEv:                     // @_ZN7myshape4moveEv
        .cfi_startproc
// %bb.0:                               // %entry
        .cfi_b_key_frame
        pacibsp
        stp     x30, x19, [sp, #-16]!           // 16-byte Folded Spill
        .cfi_negate_ra_state
        .cfi_def_cfa_offset 16
        .cfi_offset w19, -8
        .cfi_offset w30, -16
        mov     x19, x0
        ldr     x0, [x0]
        ldr     x16, [x0]
        mov     x17, x0
        movk    x17, #50297, lsl #48
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_0
        brk     #0xc472
.Lauth_success_0:
        ldr     x9, [x16]
        mov     x8, x16
        mov     x17, x8
        movk    x17, #36564, lsl #48
        blraa   x9, x17
        ldr     x0, [x19, #8]
        ldr     x16, [x0]
        mov     x17, x0
        movk    x17, #50297, lsl #48
        autda   x16, x17
        mov     x17, x16
        xpacd   x17
        cmp     x16, x17
        b.eq    .Lauth_success_1
        brk     #0xc472
.Lauth_success_1:
        ldr     x2, [x16]
        mov     x1, x16
        ldp     x30, x19, [sp], #16             // 16-byte Folded Reload
        autibsp
        eor     x16, x30, x30, lsl #1
        tbz     x16, #62, .Lauth_success_2
        brk     #0xc471
.Lauth_success_2:
        mov     x16, x1
        movk    x16, #36564, lsl #48
        braa    x2, x16
```
</spoiler>

**Review by kovdan01 — COMMENTED** [review]
With latest update, test-suite passes, thanks!

**Review by kovdan01 — COMMENTED** [review]
@atrosinenko Could you please add a test for the case you've fixed in your latest commit edaae6a75879a38ffe8cd4e064aa9aca77126ef3? I might be missing smth, but the fix looks untested.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -107,6 +107,19 @@ unsigned AArch64InstrInfo::getInstSizeInBytes(const MachineInstr &MI) const {
...
   const MCInstrDesc &Desc = MI.getDesc();
 
+  if (!MI.isBundle() && isTailCallReturnInst(MI)) {
+    NumBytes = Desc.getSize() ? Desc.getSize() : 4;
+
+    const auto *MFI = MF->getInfo<AArch64FunctionInfo>();
+    if (!MFI->shouldSignReturnAddress(MF))
+      return NumBytes;
+
+    auto &STI = MF->getSubtarget<AArch64Subtarget>();
```
[the remark this answers is not in the document]
Comment (atrosinenko) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:117]: I will fix this together with a few other updates, thank you!

**Comment by atrosinenko** [comment]
For uniformity, added `AuthCheckMethod::XPAC` to the list of all possible check methods and corresponding RUN line to `llvm/test/CodeGen/AArch64/sign-return-address-tailcall.ll` (this is not a test for edaae6a75879a38ffe8cd4e064aa9aca77126ef3, though).

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 2]
```diff
@@ -1757,38 +1762,70 @@ void AArch64AsmPrinter::emitPtrauthCheckAuthenticatedValue(
...
+                       .addReg(TestedReg)
+                       .addReg(TestedReg)
+                       .addImm(1));
+    //  tbz Xscratch, #62, Lsuccess
+    EmitToStreamer(
+        MCInstBuilder(AArch64::TBZX)
+            .addReg(ScratchReg)
+            .addImm(62)
+            .addExpr(MCSymbolRefExpr::create(SuccessSym, OutContext)));
+  } else {
```
Comment (kovdan01) [inline llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1826]: Nit: alternatively, you can change the previous `else if` to just `else` and put `Method == AuthCheckMethod::HighBitsNoTBI` condition as an assert in its beginning. Feel free to ignore.

**Review by kovdan01 — APPROVED** [review]
LGTM, thanks!

**Comment by atrosinenko** [comment]
I have just rebased the PR onto current `main` branch (the first three commits are exactly the same, except for conflict resolution with 5192cb772ad58af4b557539791ff8de60ab450a3) and added a few improvements, sorry for the delay.

According to code size metrics collected by building llvm-test-suite at different optimization levels, restricting `$AddrDisc` operand's register class actually regressed the code size. Restricting the call target, on the other hand, turned out to have negligible impact on `.text` section size (and only in a couple of tests). Thus, I updated the definition of `AUTH_TCRETURN` to avoid regressing the code size.

Presently, the PR still restricts the register class of `$AddrDisc` of `AUTH_TCRETURN_BTI`, which could still regress the code size when BTI is enabled, though the original implementation in `AArch64PointerAuth` pass turned out to be equally affected by this bug, so restricting `$AddrDisc` makes it correct at least. I added a test that catches the bug both for `AUTH_TCRETURN` and `AUTH_TCRETURN_BTI`.

@kovdan01 Are you OK with the changes?

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp (hunk 2 above)
Comment (atrosinenko) [reply llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1847]: On the one hand, a dangling 
```cpp
  } else {
    llvm_unreachable("Unsupported check method");
  }
```
block looks a bit clumsy, on the other hand, it looks natural from the standpoint of introducing completely new check methods (if any) by inserting one more `if else (... ) { ... }` block of text, so let's keep as-is for now...

File: llvm/test/CodeGen/AArch64/ptrauth-tail-call-regalloc.ll [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
+
+; Test that expansion of AUTH_TCRETURN does not crash due to unavailability of
+; neither x16 nor x17 as a scratch register.
+define i32 @test_scratch_reg_nobti(ptr %callee, ptr %addr) #0 {
+entry:
+  ; Force spilling of LR
+  tail call void asm sideeffect "", "~{lr}"()
+  ; Clobber x0-x15 and x18-x29
```
Comment (kovdan01) [inline llvm/test/CodeGen/AArch64/ptrauth-tail-call-regalloc.ll:18]: I'm not sure if I get the reason why x16 and x17 should not be clobbered for this test. Could you please clarify why?

BTW, when I locally tried to add them to clobber list, I got `braa x18, x17`, where both x18 and x17 are not spilled/restored while being temporary non-callee-saved clobbered registers (see https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#general-purpose-registers) - this does not look correct, but I might be missing smth.

**Review by kovdan01 — COMMENTED** [review]
I left several comments with questions and suggestions. Apart of them, the PR LGTM.

File: llvm/test/CodeGen/AArch64/ptrauth-tail-call-regalloc.ll (hunk 3 above)
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/ptrauth-tail-call-regalloc.ll]: Thanks for updating comments and tests!

> I guess we observe the case "... then it should use r18 as an additional temporary register", so it is probably handled the same way as x9-x15.

Yes, it's probably this case. But, even though x18 is handled as a temporary register in this case, it does not look correct to have `mov x17, x0; mov x18, x1; /* clobber x17 and x18 */; braa x18, x17`, doesn't it? I also might be missing smth and would be glad if you explain that. To achieve such assembly output, just add x16 and x17 to clobber list of the last `tail call void asm sideeffect` instruction.

File: llvm/test/CodeGen/AArch64/ptrauth-call.ll [hunk 4]
```diff
@@ -169,13 +169,13 @@ define i32 @test_tailcall_ib_var(ptr %arg0, ptr %arg1) #0 {
...
 define void @test_tailcall_omit_mov_x16_x16(ptr %objptr) #0 {
 ; CHECK-LABEL: test_tailcall_omit_mov_x16_x16:
-; CHECK:         ldr     x16, [x0]
-; CHECK:         mov     x17, x0
-; CHECK:         movk    x17, #6503, lsl #48
-; CHECK:         autda   x16, x17
-; CHECK:         ldr     x2, [x16]
-; CHECK:         movk    x16, #54167, lsl #48
-; CHECK:         braa    x2, x16
+; CHECK-NEXT:    ldr     x16, [x0]
```
[the remark this answers is not in the document]
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/ptrauth-call.ll:172]: Thanks for explanation, it just was unclear that there were trivial movs present which were not caught due to lack of `-NEXT` in check lines. Now I get it.

**Review by kovdan01 — APPROVED** [review]
LGTM

**Pull request merged**